src/main/java/org/truffleruby/language/exceptions/RescueNodeWrapper.java

```diff
@@ -50,30 +50,50 @@ public NodeCost getCost() {
             return NodeCost.NONE;
         }
 
-        @SuppressWarnings("deprecation")
         @Override
         public Object execute(VirtualFrame frame) {
-            try {
-                probeNode.onEnter(frame);
-                Object returnValue = delegateNode.execute(frame);
-                probeNode.onReturnValue(frame, returnValue);
-                return returnValue;
-            } catch (Throwable t) {
-                probeNode.onReturnExceptional(frame, t);
-                throw t;
+            Object returnValue;
+            for (;;) {
+                boolean wasOnReturnExecuted = false;
+                try {
+                    probeNode.onEnter(frame);
+                    returnValue = delegateNode.execute(frame);
+                    wasOnReturnExecuted = true;
+                    probeNode.onReturnValue(frame, returnValue);
+                    break;
+                } catch (Throwable t) {
+                    Object result = probeNode.onReturnExceptionalOrUnwind(frame, t, wasOnReturnExecuted);
+                    if (result == ProbeNode.UNWIND_ACTION_REENTER) {
+                        continue;
+                    } else if (result != null) {
+                        returnValue = result;
+                        break;
+                    }
+                    throw t;
+                }
             }
+            return returnValue;
         }
 
-        @SuppressWarnings("deprecation")
         @Override
         public void doExecuteVoid(VirtualFrame frame) {
-            try {
-                probeNode.onEnter(frame);
-                delegateNode.doExecuteVoid(frame);
-                probeNode.onReturnValue(frame, null);
-            } catch (Throwable t) {
-                probeNode.onReturnExceptional(frame, t);
-                throw t;
+            for (;;) {
+                boolean wasOnReturnExecuted = false;
+                try {
+                    probeNode.onEnter(frame);
+                    delegateNode.doExecuteVoid(frame);
```

Comment:
Shouldn't this code be shared (void and non-void). This `doExecuteVOID` is the only difference (and the final return, of course).

Reply:
I'm not really sure how to share it - the innermost code is different and using something like a lambda or a base class seems overkill.